mempool/clist_mempool_test.go

```diff
@@ -805,6 +785,211 @@ func TestMempoolNotifyTxsAvailable(t *testing.T) {
+	mockClient.On("CheckTxAsync", mock.Anything, mock.Anything).Run(func(_ mock.Arguments) {
+		reqRes := newReqRes(txs[2], 1, abci.CHECK_TX_TYPE_RECHECK)
+		callback(reqRes.Request, reqRes.Response)
+	}).Return(nil, nil).Once()
+
+	// On the fourth tx the app returns an error, so the callback will not be called by the client.
+	mockClient.On("CheckTxAsync", mock.Anything, mock.Anything).Return(nil, ErrCheckTxAsync{}).Once()
+
+	// Check that recheck has not started yet.
+	require.True(t, mp.recheck.done())
+	require.Nil(t, mp.recheck.cursor)
+	require.Nil(t, mp.recheck.end)
+
+	// mp.recheck.done() should be true only before and after calling recheckTxs.
+	mp.recheckTxs()
+	require.True(t, mp.recheck.done())
+	require.Nil(t, mp.recheck.cursor)
+	require.NotNil(t, mp.recheck.end)
+	require.Equal(t, mp.recheck.end, mp.txs.Back())
+	require.Equal(t, len(txs)-1, mp.Size()) // one invalid tx was removed
+	require.Equal(t, int32(2), mp.recheck.numPendingTxs.Load())
+
+	mockClient.AssertExpectations(t)
+}
+
+// Test that rechecking finishes correctly when a request fails, using an async ABCI client.
+func TestMempoolAsyncRecheckTxReturnError(t *testing.T) {
+	var callback abciclient.Callback
+	mockClient := new(abciclimocks.Client)
+	mockClient.On("Start").Return(nil)
+	mockClient.On("SetLogger", mock.Anything)
+	mockClient.On("Error").Return(nil).Times(4)
+	mockClient.On("SetResponseCallback", mock.MatchedBy(func(cb abciclient.Callback) bool { callback = cb; return true }))
+
+	// Mocking the async client will just send the request and not call the callback, which we do
+	// manually later. The first 4 times CheckTxAsync is called are for adding the transactions to
+	// the mempool. The other 4 are for rechecking, and in particular, it fails on the second and
+	// last transactions.
+	mockClient.On("CheckTxAsync", mock.Anything, mock.Anything).Return(nil, nil).Times(6)
+	mockClient.On("CheckTxAsync", mock.Anything, mock.Anything).Return(nil, ErrCheckTxAsync{}).Times(2)
+
+	mp, cleanup := newMempoolWithAppMock(mockClient)
+	defer cleanup()
+	mp.EnableTxsAvailable()
+
+	// First send CheckTx requests for a few txs.
+	txs := []types.Tx{[]byte{0x01}, []byte{0x02}, []byte{0x03}, []byte{0x04}}
+	for _, tx := range txs {
+		_, err := mp.CheckTx(tx)
+		require.NoError(t, err)
+	}
+	require.Zero(t, mp.Size())
+
+	// Invoke CheckTx callbacks asynchronously to add txs to mempool.
+	for _, tx := range txs {
+		reqRes := newReqRes(tx, abci.CodeTypeOK, abci.CHECK_TX_TYPE_CHECK)
+		callback(reqRes.Request, reqRes.Response)
+	}
+	require.Len(t, txs, mp.Size())
+
+	// Check that recheck has not started yet.
+	require.True(t, mp.recheck.done())
+	require.Nil(t, mp.recheck.cursor)
+	require.Nil(t, mp.recheck.end)
+
+	// The first tx is valid, the third is invalid, and the request for the second and fourth tx
+	// were not sent, so the callback is not called.
+	mockClient.On("Flush", mock.Anything).Run(func(_ mock.Arguments) {
```

Comment:
Why we need to mock the flush? I why it has to return content?

Reply:
This code is not really easy to understand.

Reply:
On the socket client, CheckTxAsync does not send the request, it just adds it to a queue. It's the Flush method that actually sends it and invokes the callback.

Reply:
I added more comments to improve clarity of the code.